pkg/functions/builder.go

```diff
@@ -51,78 +72,40 @@ func (ib *DockerImageBuilder) BuildImage(faas, fnID string, exec *Exec) (string,
 	if err != nil {
 		return "", errors.Wrap(err, "failed to create a temp dir")
 	}
-	defer cleanup(tmpDir)
+	defer os.RemoveAll(tmpDir)
 
 	log.Debugf("Created tmpDir: %s", tmpDir)
-	log.Printf("Pulling image: %s", exec.Image)
 
 	if err := images.DockerError(ib.docker.ImagePull(context.Background(), exec.Image, types.ImagePullOptions{})); err != nil {
-		return "", errors.Wrap(err, "failed to pull image")
+		return "", errors.Wrapf(err, "failed to pull image '%s'", exec.Image)
+	}
+
+	if err := ib.copyFunctionTemplate(tmpDir, exec.Image); err != nil {
+		return "", err
 	}
 
-	if err := writeFunctionDockerfile(tmpDir, ib.functionTemplateDir, faas, exec); err != nil {
+	if err := writeFunctionFile(tmpDir, exec); err != nil {
 		return "", errors.Wrap(err, "failed to write dockerfile")
 	}
 
-	err = images.BuildAndPushFromDir(ib.docker, tmpDir, name, ib.registryAuth)
+	buildArgs := map[string]*string{
+		"IMAGE":        ptr(exec.Image),
+		"FUNCTION_SRC": ptr(functionFile),
+	}
+	err = images.BuildAndPushFromDir(ib.docker, tmpDir, name, ib.registryAuth, buildArgs)
 	return name, err
 }
 
-func cleanup(tmpDir string) {
-	defer trace.Tracef("rm dir: %s", tmpDir)()
-	os.RemoveAll(tmpDir)
+const functionFile = "function.txt"
+
+func ptr(s string) *string {
```

Comment:
or swag.String(...) :)

Reply:
Sure! Although, in Go, reuse seems to be an anti-pattern :)